packages/block-library/src/navigation/edit/unsaved-inner-blocks.js

```diff
@@ -33,13 +49,41 @@ export default function UnsavedInnerBlocks( {
 	onSave,
 	hasSelection,
 } ) {
+	const originalBlocks = useRef();
+
+	// Initially store the uncontrolled inner blocks for
+	// dirty state comparison.
+	if ( ! originalBlocks?.current ) {
```

Comment:
Is modifying the ref directly like this advised? Would it be better to put the ref directly on the component?

Reply:
What makes you concerned here?

The ref is a stored value which we create solely to store whatever blocks the component is passed as its initial value.

We do nothing with `originalBlocks` in terms of passing it to `<InnerBlocks>`. Its sole purpose is to act as a source of truth for the original "state" of the blocks that was originally passed to the component. 

I'm going to double check this checks out correctly and I'm not setting us up for bugs.

Reply:
I guess my concern is whether we are abusing the `ref` feature of React - why not just us a local variable?

Reply:
Because a ref is guaranteed to be the same reference for the _lifetime_ of the component whereas the local variable will be a new reference on each render of the component:

> [The returned object will persist for the full lifetime of the component](https://reactjs.org/docs/hooks-reference.html#useref) (React Docs)

Reply:
So basically I'm using the ref to store the "source of truth" for the lifecycle of the component. Then if at any point the current blocks don't match up with the original source of truth reference stored in the `ref` then the blocks have changed because the references are different.

Reply:
Would it be possible to replace this ref with a combination of `hasUncontrolledInnerBlocks` and `hasSavedUnsavedInnerBlocks` that's used in edit/index.js?

Reply:
Basically I wonder if this information can be derived from somewhere instead of computing it here

Reply:
Wow, great exploration @getdave 👍  alright, let's stick to the ref then.